tests/unit/test_units.py

```diff
@@ -0,0 +1,48 @@
+import numpy as np
+import numpy.testing as npt
+import pytest
+from scmdata.testing import get_single_ts
+
+from local.h2_adjust.units import _get_unit_scaling, sanitize_production_intensity_units
+
+
+@pytest.mark.parametrize(
+    "source,target,exp",
+    [
+        ("kg NOx / MWh", "kg NOx / MWh", 1),
+        ("kg NOx / MWh", "t NOx / MWh", 1 / 1000),
+        ("g NOx/kWh", "kg NOx / MWh", 1),
+        ("kg NOx/TJ", "kg NOx / MWh", 3600 / 1e6),
+        ("kg /t", "kg / kg", 1 / 1000),
+    ],
+)
+def test_convert_intensities(source, target, exp):
+    np.testing.assert_almost_equal(_get_unit_scaling(source, target), exp)
+
+
+@pytest.mark.parametrize(
+    "unit,carrier,product,exp",
+    (
+        ("% of H2", "H2", "H2", 0.01),
+        ("kgNOx/tNH3", "NH3", "NOx", 1 / 1000 * (14 + 3) / 3),
+        ("kgNH3/tNH3", "NH3", "NH3", 1 / 1000 * (14 + 3) / 3),
+        ("% H2 component of fuel", "NH3", "H2", 0.01),
+        ("% H2 component of fuel", "CH4", "H2", 0.01),
+        ("% of fuel (LNG) consumption", "CH4", "CH4", 0.01 * (12 + 4) / 4),
+        ("% NH3 used", "NH3", "NH3", 0.01 * (14 + 3) / 3),
```

Comment:
The by mass thing here is surprising me too. So, we're saying, if we end up using 1 kg NH3 in some final end use, 0.01 * (14 + 3) / 3 kg of H2 was required to produce that? Having it be by mass is again something I would check (by volume or by amount would have made more sense to me, surely there is some chemical reaction which dictates what the ratios should be). Or does this come from our assumptions around leakage i.e. we're blending in an assumption about how much leakage (that's where the % comes from) and an assumption about the mass ratios between the things? (In which case, it seems odd to me that the stuff leaks as NH3 or CH4 rather than as H2 but maybe that's the simplest way to get a ballpark figure...)

Reply:
This is leakage (slip in this case from incomplete combustion). 0.021% of NH3 produced will be emitted at the point of combustion.

These conversions are used to convert the amount of H2 produced for a given carrier/sector into the mass of leakage emissions. We are assuming that 1 kg of H2 is equivalent (by mass) to 17/3 kg of NH3 and that conversion is lossless. That is the upper bound of this conversion as there is obviously losses involved and we should mention that in the manuscript

Reply:
So this is two steps in one? It's both the H2 -> NH3 step (that's where (14 + 3) / 3 comes from) and the leakage factor (that's where the 0.01 i.e. 1% comes from)?

Reply:
Put another way, I would have assumed we'd just assume that 1% of CH4 and 1% of NH3 would leak. The way this is written 0.021% of NH3 leaks while 0.025% of CH4 leaks (and I can't work out why we would assume the two are different just because of their different masses)

Reply:
> So this is two steps in one?

Yes. We are deriving the mass of NH3 per kg of H2 produced. It is the "per kg of H2" that causes the need to do the mass conversion even though the product and the carrier are both the same.

Reply:
Ah ok right yes that context is important. The input units is % NH3 but we want to go to per kg of H2 hence mass factor matters. My bad again, trying to do this without full context in my head biting again (maybe comment in function to help explain is again best we can do...)

Reply:
Yes, these unit derivations are hard as each is ever so slightly different. At least it is all in one location now so from the 2nd notebook on in the processing chain we have a consistent set of factors to apply.